### PR TITLE
Better modularisation in `icu_provider`

### PR DESCRIPTION
**Pull request by robertbastian**
This PR namesspaces all macros in `icu_provider`, and exports every type only under a single path (as well as the prelude).

### CODE REVIEW

**Comment by robertbastian** [comment]
Before: https://unicode-org.github.io/icu4x/rustdoc/icu_provider/index.html#reexports

After: https://icu4x-pr-artifacts.storage.googleapis.com/gha/99198f0e60e6f134b4ab73803f88f8f739d6f7f6/rustdoc/icu_provider/index.html#modules

File: ffi/capi/src/provider.rs [hunk 1]
```diff
@@ -4,9 +4,9 @@
 
 #[allow(unused_imports)] // feature-specific
 use alloc::boxed::Box;
-use icu_provider::prelude::*;
 #[allow(unused_imports)] // feature-specific
-use icu_provider::MaybeSendSync;
+use icu_provider::any::MaybeSendSync;
```
Comment (sffc) [inline ffi/capi/src/provider.rs:8]: Issue: This trait isn't really `any` related and should probably be in a different module

File: ffi/capi/src/provider.rs [hunk 2]
```diff
@@ -297,7 +297,7 @@ macro_rules! call_constructor_unstable {
             ))?,
             $crate::provider::ICU4XDataProviderInner::Empty => $unstable(&icu_provider_adapters::empty::EmptyDataProvider::new(), $($args,)*),
             #[cfg(feature = "buffer_provider")]
-            $crate::provider::ICU4XDataProviderInner::Buffer(buffer_provider) => $unstable(&icu_provider::AsDeserializingBufferProvider::as_deserializing(buffer_provider), $($args,)*),
+            $crate::provider::ICU4XDataProviderInner::Buffer(buffer_provider) => $unstable(&icu_provider::serde::AsDeserializingBufferProvider::as_deserializing(buffer_provider), $($args,)*),
```
Comment (sffc) [inline ffi/capi/src/provider.rs:300]: Thought/Suggestion: merge `serde` and `buf`

**Review by sffc — COMMENTED** [review]
Not convinced this makes things better; I understand the desire to reduce re-exports but then why not make the modules more private? At least the macros are convenient to have on the top level since they are usually called in a chain, whereas the other types can be imported.

File: ffi/capi/src/provider.rs (hunk 2 above)
Comment (robertbastian) [reply ffi/capi/src/provider.rs:300]: Happy to, I thought that as well.

File: ffi/capi/src/provider.rs (hunk 1 above)
Comment (robertbastian) [reply ffi/capi/src/provider.rs:8]: It's only used for bounds in this module, so the only usages outside this module are `AnyPayloadProvider` and one tutorial that uses any.

**Review by sffc — APPROVED** [review]

**Pull request merged**